recipes/libmp3lame/all/conanfile.py

```diff
@@ -47,121 +40,123 @@ def _user_info_build(self):
         return getattr(self, "user_info_build", self.deps_user_info)
 
     def export_sources(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            self.copy(patch["patch_file"])
+        export_conandata_patches(self)
 
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
 
     def configure(self):
         if self.options.shared:
-            del self.options.fPIC
-        del self.settings.compiler.libcxx
-        del self.settings.compiler.cppstd
+            try:
+                del self.options.fPIC
+            except Exception:
+                pass
+        try:
+            del self.settings.compiler.libcxx
+        except Exception:
+            pass
+        try:
+            del self.settings.compiler.cppstd
+        except Exception:
+            pass
+
+    def layout(self):
+        basic_layout(self, src_folder="src")
 
     def build_requirements(self):
-        if not self._is_msvc and not self._is_clang_cl:
+        if not is_msvc(self) and not self._is_clang_cl:
             self.build_requires("gnu-config/cci.20201022")
-            if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
-                self.build_requires("msys2/cci.latest")
+            if self.settings.os == "Windows":
+                self.win_bash = True
+                if not self.conf.get("tools.microsoft.bash:path", default=False, check_type=str):
+                    self.tool_requires("msys2/cci.latest")
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version],
-                  destination=self._source_subfolder, strip_root=True)
+        get(self, **self.conan_data["sources"][self.version], destination=self.source_folder, strip_root=True)
 
-    def _apply_patch(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
-        tools.replace_in_file(os.path.join(self._source_subfolder, "include", "libmp3lame.sym"), "lame_init_old\n", "")
 
-    @contextmanager
-    def _msvc_build_environment(self):
-        with tools.chdir(self._source_subfolder):
-            with tools.vcvars(self.settings):
-                with tools.environment_append(VisualStudioBuildEnvironment(self).vars):
-                    yield
+    def _generate_vs(self):
+        tc = VCVars(self)
+        tc.generate()
 
     def _build_vs(self):
-        with self._msvc_build_environment():
-            shutil.copy("configMS.h", "config.h")
+        with chdir(self, self.source_folder):
+            shutil.copy2("configMS.h", "config.h")
             # Honor vc runtime
-            tools.replace_in_file("Makefile.MSVC", "CC_OPTS = $(CC_OPTS) /MT", "")
+            replace_in_file(self, "Makefile.MSVC", "CC_OPTS = $(CC_OPTS) /MT", "")
```

Comment:
It's a lie now, because there is no more injection of runtime through `VisualStudioBuildEnvironment`.

See one of the logs of MT static test package: https://c3i.jfrog.io/c3i/misc/logs/pr/13791/5-configs/windows-visual_studio/libmp3lame/3.100//6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7-test.txt

```
icrosoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building Custom Rule C:/J/w/prod/BuildSingleReference@2/conan-center-index/recipes/libmp3lame/all/test_package/CMakeLists.txt
  test_package.c
LINK : warning LNK4098: defaultlib 'LIBCMT' conflicts with use of other libs; use /NODEFAULTLIB:library [C:\J\w\prod\BuildSingleReference@2\conan-center-index\recipes\libmp3lame\all\test_package\build\test_package.vcxproj]
  test_package.vcxproj -> C:\J\w\prod\BuildSingleReference@2\conan-center-index\recipes\libmp3lame\all\test_package\build\Release\test_package.exe
  Building Custom Rule C:/J/w/prod/BuildSingleReference@2/conan-center-index/recipes/libmp3lame/all/test_package/CMakeLists.txt
libmp3lame/3.100 (test package): Running test()
```

Reply:
hrm, this is a big deal that I wouldn't have realised was a problem until I tried to link my final main project ... thanks @SpaceIm